### PR TITLE
Clickhouse support

### PR DESCRIPTION
**Pull request by tsl-karlp**
Beginnings of ClickHouse driver.  Thanks to @b.mukvich for the initial work!

I had trouble running the tests.  Not sure how to run just these new tests.

### CODE REVIEW

**Comment by camsaul** [comment]
@tsl-karlp WRT to this error

```
# env ENGINES=h2,clickhouse lein test
... snip....
INSERT FAILED:
["INSERT INTO venues (name, latitude, longitude, price, category_id) VALUES (?, 10.0646, -165.374, 3, 4)" "Red Medicine"]

ClickHouseException:
 Message: ClickHouse exception, code: 27, host: localhost, port: 8123; Code: 27, e.displayText() = DB::Exception: Cannot parse input: expected \t before: \n: (at row 1)

Row 1:
Column 0,   name: name,        type: String,  parsed text: "Red Medicine"
ERROR: Line feed found where tab is expected. It's like your file has less columns than expected.
And if your file have right number of columns, maybe it have unescaped backslash in value before tab, which cause tab has escaped.

, e.what() = DB::Exception

 SQLState: null
 Error Code: 27
```

I Googled that error really quick and found this: https://github.com/yandex/ClickHouse/issues/2652. So my guess is it seems like it's trying to insert your rows in CSV format. So perhaps start by looking for the reason it's trying to insert them that way instead of the normal SQL syntax

File: src/metabase/db/spec.clj [hunk 1]
```diff
@@ -26,6 +26,19 @@
...
+   You can also optionally set host and port."
+  [{:keys [host port db]
+    :or {host "localhost", port 8123, db "default"}
+    :as opts}]
+  (merge {:classname "ru.yandex.clickhouse.ClickHouseDriver" ; must be in classpath
+          :subprotocol "clickhouse"
+          ; :subname (str "//" host ":" port "/" db)}
+          :subname (str "//" host ":" port)}
+         (dissoc opts :host :port :db)))
+
```
Comment (camsaul) [inline src/metabase/db/spec.clj]: Read the message at the end of this file 😉

**Comment by tsl-karlp** [comment]
Thanks, @camsaul !

I don't think Clickhouse itself supports the normal SQL syntax for insertion.  The clickhouse-jdbc wrapper, when using PreparedStatements, will POST something like:

```
INSERT INTO test FORMAT CSVWithNames
column1, column2, column3
1, 1, 1
2, 2, 2
```

Can the tests be modified to use PreaparedStatements, or should the Clickhouse tests have their own way of inserting the data?

**Comment by camsaul** [comment]
Yeah it seems you might need to write a custom implementation of `load-data!` for Clickhouse so you can insert data in a way it likes

**Comment by tsl-karlp** [comment]
@camsaul : 
> Yeah it seems you might need to write a custom implementation of load-data! for Clickhouse so you can insert data in a way it likes

Is there a good example of doing this in the code?

**Comment by camsaul** [comment]
Not really, all of the other SQL drivers support normal SQL INSERT syntax in some form or another. But It shouldn't be too hard.

**Comment by Slach** [comment]
guys, thanks for your great work

**Comment by ei-grad** [comment]
I have got my hands on this, and found that the latest code in twosixlabs/metabase@acbb5abdf8f020558c42e47fb292b0c3016fc45b doesn't compile because `db.spec/clickhouse` is used in `metabase.driver.clickhouse/connection-details->spec`. Fixed, squashed/rebased on current master, added some (mostly cosmetic) stuff. Tests still needs a `load-data`, and it still throws the NullPointerException with unreadable stacktrace when trying to execute a query via metabase web UI. Working on it.

**Comment by Slach** [comment]
@ei-grad thank you for your work!

**Comment by tlrobinson** [comment]
Hey all,

Thank you for this PR and apologies for the long delay responding. We've recently done some work to support drivers as plugins in Metabase and as of the next release (v0.32) we will be asking driver developers to first publish new drivers in their own repositories (typically named `metabase-*-driver`). If the driver fully passes the driver tests we will link to it from our documentation, and if a driver gains significant usage we will consider merging it into Metabase itself.

Here is some documentation about publishing a driver as a plugin in a separate repository: https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics#drivers-shipped-as-3rd-party-plugins

Here are two example drivers: https://github.com/metabase/sudoku-driver https://github.com/metabase/crate-driver

We still have some work to do on the documentation. You can follow progress and provide feedback on that here: https://github.com/metabase/metabase/issues/9348

Thank you.

**Comment by tlrobinson** [comment]
By the way, I noticed there's another Clickhouse driver: #9469 You all may want to consider consolidating efforts.

**Pull request closed without merging**